### ISSUE TITLE
Propagate `adjoint self` information to C# code gen

### ISSUE THREAD

**Issue by msoeken** [microsoft/qsharp-runtime#473]
** Please describe what you would like the feature to accomplish.**
Runtime metadata sets `IsAdjoint` to `true` even for self-adjoint operations:
https://github.com/microsoft/qsharp-runtime/blob/928067442380524ec3e63f8812403364561478c5/src/Simulation/Core/Operations/Adjoint.cs#L78

We recently discovered this to be a problem when modifying the implementation of SWAP (microsoft/qsharp-runtime#458) using conjugation, which lead to one of the CNOTs be self-adjoint in IQ#'s execution path trace (microsoft/iqsharp#392).

** Describe the solution you'd like **
The `IsAdjoint` runtime metadata should be false for `Adjoint` calls to self-adjoint operations.

** Describe alternatives you've considered **
So far we are ignoring this information and assume it explicitly for known self-adjoint operations.

**Comment by cgranade** [comment]
To clarify, this is specific to the runtime metadata used by the execution path tracer?

**Comment by msoeken** [comment]
Not necessarily. So far the method `GetRuntimeMetadata` is only used by execution path tracer, as far as I know. But it could be used from other API as well. This issue does not address fixing the problem only for execution path tracer.
